### PR TITLE
Appointment Block | Route Error Handling | Error Page

### LINKED ISSUE #26: Create appointment_blocks helper methods
Create functions to allow for TAs to make appointment blocks so that they can give default times they are available

### PR DESCRIPTION
**Pull request by DylanHalstead**
_I got tired of working on #26 for so long I did a couple other things as well in this PR, sorry_

Closes #26 

# Problem
## Appointment Blocks
Members of a sections Instructional Team need a way to input their office hour availability in the DB

## Route Error Handling
We need to check certain things on certain routes (is the user signed in, is the user a part of the section they're trying to view, are the a part of the instructional team, etc...)

## Error Page
We need a user-friendly way of showing users errors and allowing them to return to a safe place.

# Solution
## Appointment Blocks
Members of the instructional team can now submit their availability at `/courses/[courseID]/sections/[sectionID]/members/[memberID]/blocks/create`. To visit this page, you must be signed in as the user with this respective `memberID` and must be a member of the instructional team. The form has both client and server-side error handling:
image
image
image
image

On the server end, you cannot submit a block that conflicts with another (if they intersect). If you attempt to create a block that starts when another block ends (or in between), the block(s) are deleted and concatenated into one big block (to make displaying on the calendar easier). This has the byproduct of possibly deleting already made appointments, but maybe we can add some middleware to _transfer_ the appointments to the new block if we think this is necessary.
image

## Route Error Handling
Routes that require a user to be logged in are now under the `signed_in` folder with a respective `+layout.server.ts` checking such. There are 2 other layouts made which are self-explanatory

## Error Page
Made a standard Error page that shows them the status and message
image

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  const [hours, minutes] = time.split(':').map(num => parseInt(num));
+
+  const date = new Date(0);
+  date.setHours(hours-parseInt(timeZoneOffset));
+  date.setMinutes(minutes);
+
+  return date;
+}
+
+export function stringToWeekDay(dayString: string): WeekDay | undefined {
```
Comment (DylanHalstead) [inline src/lib/utils.ts:19]: There may be a better way to do this with some kinda of `Objects.value(WeekDay)`

File: src/routes/(signed_in)/courses/[courseID]/sections/[sectionID]/members/[memberID]/blocks/create/+page.svelte [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+					class="input input-bordered w-full max-w-xs"
+					type="time"
+					name="end-time"
+					id="end-time"
+					bind:value={endTime}
+					required
+				/>
+			</div>
+			{#if startTime !== '' && endTime !== '' && !areDatesValid}
+				<p class="text-red-500">End time must be after start time</p>
```
Comment (DylanHalstead) [inline src/routes/(signed_in)/courses/[courseID]/sections/[sectionID]/members/[memberID]/blocks/create/+page.svelte:70]: This is kinda jittery atm; May want to add some cleaner CSS here later

File: src/lib/utils.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+export function postgresTimeWithTimeZoneToDate(timeWithZone: string): Date {
+	const [time, timeZoneOffset] = timeWithZone.split("-");
+	if (!time || !timeZoneOffset) {
+		console.error('Invalid input format. Expected "HH:MM-TZ".');
+	}
+
+	const [hours, minutes] = time.split(":").map(num => parseInt(num));
+
+	const date = new Date(0);
+	date.setHours(hours - parseInt(timeZoneOffset));
```
Comment (DylanHalstead) [inline src/lib/utils.ts:13]: This feels jank. Likely better way to do this if anyone knows

**Review by kfinberg7581 — APPROVED** [review]
Here is your upvote!

LGTM ty for hard work

File: src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    const usersMembers = await getUsersSectionMembers(userID);
+    if (!usersMembers.some(member => member.id === params.memberID)) {
+      error(403,{ message: 'Forbidden: Cannot make block for other members' })
+    }
+    // Validate the user is part of instructional team for the section
+    const sectionMembers = await getSectionsSectionMembers(params.sectionID);
+    if (!sectionMembers.some(member => member.user_id === userID && (member.member_type === 'ta' || member.member_type === 'instructor'))) {
+      error(403,{ message: 'Forbidden: Not a part of the instructional team' })
+    }
+    // Validate the input
```
Comment (jadenPete) [inline src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts]: You might be able to validate more easily (and safely) if you pull in a library like:
https://github.com/sinclairzx81/typebox
https://github.com/gcanti/io-ts

File: src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    if (!dayofWeek || !startTime || !endTime ) {
+      error(400,{ message: 'Invalid input: missing input' })
+    }
+
+    if (!Object.values(WeekDay).includes(dayofWeek)) {
+      error(400,{ message: 'Invalid input: Invalid day of week' })
+    }
+
+    // convert the start-time and end-time to Date objects
+    const start = new Date(0);
```
Comment (jadenPete) [inline src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts]: Could we move this into a function in `utils.ts`?

**Review by jadenPete — APPROVED** [review]

File: src/lib/utils.ts (hunk 1 above)
Comment (jadenPete) [reply src/lib/utils.ts:19]: https://www.typescriptlang.org/play?#code/KYOwrgtgBAwg9nANgBWAJwM5xFA3gKCiKgCkBDAE1CgF4oAiAK0tHoBpDjyJgNaHmPDO05EA0mmAA3MgBd+9ANaSZs+vgC++fADMwIAMayAltigBzYLICi4CADUyiMMAA8AQSjAAHrNAU+ACVgAzg0ClcMWTRjEHM2KCiYuIA+FIAKGWdgAC5E6Nj4rzsAfTz3AEpygG1FYABPOB0odwBdKAAfKH0qHVjgCjxRKB0wqHTQkCioOvqoWOLIEoqh4jX55vTQJdqG9po6LJcVgnWzqElZMDQcbYgS3frWgG5hta137TXL65we4D6IAGry0+EmWEQwAAdIg4OZ0pYbHZHNl0vQKPVEGQQOxYAgUOgsCAKhVXuCkNDYfDEbZICiXGjlNI5Lj4EhUJhsCTnkA

**Review by jadenPete — APPROVED** [review]

File: src/lib/utils.ts (hunk 3 above)
Comment (jadenPete) [reply src/lib/utils.ts:13]: 🤷

File: src/lib/db/appointmentBlocks.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+export async function getAppointmentBlock(id: string): Promise<AppointmentBlock>{
+  return withConnection(async (client: PoolClient) => {
+    const query: QueryConfig = {
+      text: 'SELECT appointment_blocks.id, appointment_blocks.start_time, appointment_blocks.end_time, appointment_blocks.notes FROM appointment_blocks WHERE id = $1',
+      values: [id],
+    };
+
+    const res: QueryResult<PostgresAppointmentBlock> = await client.query(query);
+    const appointmentBlocks = res.rows.map((row) => {
```
[the remark this answers is not in the document]
Comment (DylanHalstead) [reply src/lib/db/appointmentBlocks.ts]: Updated functions and `withConnection` so if for any reason an error is thrown, return undefined

File: src/lib/db/appointmentBlocks.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+export async function getMembersAppointmentBlocks(memberId: string): Promise<AppointmentBlock[]>{
+  return withConnection(async (client: PoolClient) => {
+    const query: QueryConfig = {
+      text: 'SELECT id, instructional_member_id, week_day, start_time, duration FROM appointment_blocks WHERE instructional_member_id = $1',
+      values: [memberId],
+    };
+
+    const res: QueryResult<PostgresAppointmentBlock> = await client.query(query);
+    const appointmentBlocks = res.rows.map((row) => {
+      const appointmentBlock: AppointmentBlock = {
```
[the remark this answers is not in the document]
Comment (DylanHalstead) [reply src/lib/db/appointmentBlocks.ts]: Created `postgresAppointmentBlockToAppointmentBlock`

File: src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import type { Actions } from './$types';
+import { Interval, WeekDay } from '../../../../../../../../../lib/types';
+import { getSectionsSectionMembers, getUsersSectionMembers } from '../../../../../../../../../lib/db/sectionMembers';
```
[the remark this answers is not in the document]
Comment (DylanHalstead) [reply src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts]: 🤭

File: src/lib/db/sectionMembers.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import type { QueryConfig, QueryResult, PoolClient } from 'pg';
+import { withConnection } from './index'; 
+import type { SectionMember } from "../types";
+
+export async function getUsersSectionMembers(userId: string): Promise<SectionMember[]>{
+  return withConnection(async (client: PoolClient) => {
+    const query: QueryConfig = {
+      text: 'SELECT * FROM section_members WHERE user_id = $1',
```
[the remark this answers is not in the document]
Comment (DylanHalstead) [reply src/lib/db/sectionMembers.ts]: Yeah I should've caught this. ty!

File: src/routes/+layout.server.ts [hunk 10]
```diff
@@ -27,6 +28,7 @@ export const load: LayoutServerLoad = loadFlash(async ({ locals, cookies, fetch
         setFlash(message, cookies)
       } else {
         db = {user: await createUser(locals.user.email, locals.user.given_name, locals.user.family_name)};
+        cookies.set('userID', db.user.id, {path: '/'});
```
[the remark this answers is not in the document]
Comment (DylanHalstead) [reply src/routes/+layout.server.ts]: > The httpOnly and secure options are true by default (except on [http://localhost](vscode-file://vscode-app/opt/visual-studio-code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html), where secure is false), and must be explicitly disabled if you want cookies to be readable by client-side JavaScript and/or transmitted over HTTP. The sameSite option defaults to lax.

File: src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts (hunk 5 above)
Comment (DylanHalstead) [reply src/routes/courses/[courseID]/sections/[sectionID]/members/[memberID]/appointments/create/+page.server.ts]: created `formTimeToDate`

**Pull request merged**